### PR TITLE
*: Added End-to End Style Test Coverage for LRS Scenarios in xDS Client

### PR DESCRIPTION
**Pull request by RyanBlaney**
Addresses #7704 

This pull request adds test coverage for Load Reporting Service (LRS) scenarios within the `xdsclient` package. The tests address various cases, such as stream failures, backoff mechanisms, and resource request handling when streams are unavailable.

### Highlights:

- **Test Coverage**:
  - Adds three backoff tests for LRS, modeled after the existing Aggregated Discovery Service (ADS) backoff tests.
  - Validates behavior under edge cases, including stream errors and retries.
  - Ensures proper handling of resource requests when streams are temporarily unavailable.

- **Consistency**:
  - Aligns LRS behavior testing with established ADS test patterns to maintain consistency.

### Release Notes:

- Adds test coverage for LRS stream backoff and error scenarios.

### Request for Feedback

1. **`WatchListener` Refinements**:
   - Suggestions for improving the `WatchListener` implementation to better support LRS.
   - Potential approaches for introducing a dedicated LRS-specific watcher for enhanced clarity and modularity.

2. **Test Isolation**:
   - Recommendations for isolating `load-report` test cases from broader listener tests to improve test readability and maintainability.

3. **Error Simulation**:
   - Feedback on the approach to simulating edge cases, particularly around stream errors and retries, to ensure robust test coverage.

### CODE REVIEW

File: stats/stats_test.go [hunk 1]
```diff
@@ -59,8 +59,10 @@ func init() {
 	grpc.EnableTracing = false
 }
 
-type connCtxKey struct{}
-type rpcCtxKey struct{}
+type (
```
Comment (purnesh42H) [inline stats/stats_test.go:62]: are the stats_test changes? Please revert the file if unrelated

File: xds/internal/xdsclient/tests/lrs_stream_backoff_test.go [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource/version"
+	"google.golang.org/protobuf/testing/protocmp"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	v3discoverypb "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+)
+
+// Tests the case where the management server returns an error in the ADS
```
Comment (purnesh42H) [inline xds/internal/xdsclient/tests/lrs_stream_backoff_test.go:42]: copy paste error: `ADS Stream` -> `LRS Stream`

File: xds/internal/xdsclient/tests/lrs_stream_backoff_test.go [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+// streaming RPC. Verifies that the LRS stream is restarted after a backoff
+// period, and that the previously requested resources are re-requested on the
+// new stream.
+func (s) TestLRS_BackoffAfterStreamFailure(t *testing.T) {
+	// Setup channels for state tracking.
+	streamCloseCh := make(chan struct{}, 1)
+	resourceRequestCh := make(chan []string, 1)
+	backoffCh := make(chan struct{}, 1)
+
+	// Context with timeout for the test.
```
Comment (purnesh42H) [inline xds/internal/xdsclient/tests/lrs_stream_backoff_test.go:52]: no need of this comment. Code is self explainatory

File: xds/internal/xdsclient/tests/lrs_stream_backoff_test.go [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+	}
+
+	// Create xDS client with backoff behavior.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+	client := createXDSClientWithBackoff(t, bc, streamBackoff)
+
+	// Start watching a resource.
+	const listenerName = "listener"
```
Comment (purnesh42H) [inline xds/internal/xdsclient/tests/lrs_stream_backoff_test.go:75]: Since we are writing tests for LRS stream which is used for load reporting, we don't need watchers here. What we need is for client to report load and verify if the server's channel received the LRS request.

Can you take a look at how tests are written and how the lrs server setup here https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/tests/loadreport_test.go ? This will give you some idea on how to go about it

**Comment by purnesh42H** [comment]
@RyanBlaney the tests you have written are for ADS stream which is sending discover requests. For this we need e2e tests for LRS stream which sends load requests to management server. Please write similar tests as here https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/tests/loadreport_test.go

File: xds/internal/xdsclient/tests/lrs_stream_backoff_test.go (hunk 4 above)
Comment (RyanBlaney) [reply xds/internal/xdsclient/tests/lrs_stream_backoff_test.go:75]: Okay, I had a feeling this was a mistake. Thanks, I'll get this fixed.

**Comment by purnesh42H** [comment]
@RyanBlaney are you working on this?

**Comment by RyanBlaney** [comment]
Yes I've made changes I'm just not finished yet. I've had a busy few weeks so I haven't had a lot of time to finish this.

**Comment by purnesh42H** [comment]
Thanks @RyanBlaney. I will close the PR for now. Please re-open once ready for review
